pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/SegmentLocalFSDirectory.java

```diff
@@ -0,0 +1,420 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.store;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.commons.configuration.ConfigurationException;
+import org.apache.commons.io.FileUtils;
+import org.apache.pinot.segment.spi.creator.SegmentVersion;
+import org.apache.pinot.segment.spi.index.metadata.SegmentMetadataImpl;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.segment.spi.store.ColumnIndexDirectory;
+import org.apache.pinot.segment.spi.store.ColumnIndexType;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.segment.spi.store.SegmentDirectoryPaths;
+import org.apache.pinot.spi.utils.ReadMode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.segment.spi.creator.SegmentVersion.v1;
+import static org.apache.pinot.segment.spi.creator.SegmentVersion.v2;
+import static org.apache.pinot.segment.spi.creator.SegmentVersion.v3;
+
+
+public class SegmentLocalFSDirectory extends SegmentDirectory {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentLocalFSDirectory.class);
+
+  // matches most systems
+  private static final int PAGE_SIZE_BYTES = 4096;
+  // Prefetch limit...arbitrary but related to common server memory and data size profiles
+  private static final long MAX_MMAP_PREFETCH_PAGES = 100 * 1024 * 1024 * 1024L / PAGE_SIZE_BYTES;
+  private static final double PREFETCH_SLOWDOWN_PCT = 0.67;
+  private static final AtomicLong PREFETCHED_PAGES = new AtomicLong(0);
+
+  private final File _indexDir;
+  private final File _segmentDirectory;
+  private final SegmentLock _segmentLock;
+  private SegmentMetadataImpl _segmentMetadata;
+  private final ReadMode _readMode;
+
+  private ColumnIndexDirectory _columnIndexDirectory;
+
+  public SegmentLocalFSDirectory(File directory, ReadMode readMode)
+      throws IOException {
+    this(directory, new SegmentMetadataImpl(directory), readMode);
+  }
+
+  @VisibleForTesting
+  public SegmentLocalFSDirectory(File directoryFile, SegmentMetadataImpl metadata, ReadMode readMode) {
+
+    Preconditions.checkNotNull(directoryFile);
+    Preconditions.checkNotNull(metadata);
+
+    _indexDir = directoryFile;
+    _segmentDirectory = getSegmentPath(directoryFile, metadata.getVersion());
+    Preconditions.checkState(_segmentDirectory.exists(), "Segment directory: " + directoryFile + " must exist");
+
+    _segmentLock = new SegmentLock();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when SegmentLocalFSDirectory.load throws an exception, the *SegmentLocalFSDirectory* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_segmentLock* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).